pd panic

## Bug Report

<!-- Please answer these questions before submitting your issue. Thanks! -->
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x4cc0c58]

goroutine 3078 [running]:
github.com/pingcap/pd/v4/server/schedule/storelimit.(*StoreLimit).Mode(...)
	/Users/pingcap/gopkg/src/github.com/pingcap/pd/server/schedule/storelimit/store_limit.go:127
github.com/pingcap/pd/v4/server/schedule.(*OperatorController).SetAllStoresLimitAuto(0xc00034d860, 0x3ffaaaaaaaaaaaab, 0x1)
	/Users/pingcap/gopkg/src/github.com/pingcap/pd/server/schedule/operator_controller.go:901 +0x2d8
github.com/pingcap/pd/v4/server/cluster.(*StoreLimiter).Collect(0xc001189e80, 0xc002155260)
	/Users/pingcap/gopkg/src/github.com/pingcap/pd/server/cluster/store_limiter.go:68 +0x2c9
github.com/pingcap/pd/v4/server/cluster.(*RaftCluster).HandleStoreHeartbeat(0xc0001063c0, 0xc002155260, 0x0, 0x0)
	/Users/pingcap/gopkg/src/github.com/pingcap/pd/server/cluster/cluster.go:432 +0x5dd
github.com/pingcap/pd/v4/server.(*Server).StoreHeartbeat(0xc000473c20, 0x57a4f20, 0xc002403e90, 0xc002403ec0, 0xc000473c20, 0x57b3f80, 0xc002025800)
	/Users/pingcap/gopkg/src/github.com/pingcap/pd/server/grpc_service.go:293 +0x1b2
github.com/pingcap/kvproto/pkg/pdpb._PD_StoreHeartbeat_Handler.func1(0x57a4f20, 0xc002403e90, 0x54102a0, 0xc002403ec0, 0x44d4144, 0xc002025800, 0xc002403f80, 0x3)
	/Users/pingcap/gopkg/pkg/mod/github.com/pingcap/kvproto@v0.0.0-20200409042513-05af14db7537/pkg/pdpb/pdpb.pb.go:4626 +0x89
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1(0x57a4f20, 0xc002403e90, 0x54102a0, 0xc002403ec0, 0x17, 0xc002623b30, 0x6ef08c0, 0x1)
	/Users/pingcap/gopkg/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.0.1-0.20190118093823-f849b5445de4/chain.go:31 +0x114
github.com/grpc-ecosystem/go-grpc-prometheus.(*ServerMetrics).UnaryServerInterceptor.func1(0x57a4f20, 0xc002403e90, 0x54102a0, 0xc002403ec0, 0xc00282b5a0, 0xc002623ae0, 0xc002403e30, 0xc0025f29a0, 0x6ef08c0, 0x5351c60)
	/Users/pingcap/gopkg/pkg/mod/github.com/grpc-ecosystem/go-grpc-prometheus@v1.2.0/server_metrics.go:107 +0xad
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1(0x57a4f20, 0xc002403e90, 0x54102a0, 0xc002403ec0, 0x6ef2180, 0x40515c2, 0x4064480, 0xc0017137f0)
	/Users/pingcap/gopkg/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.0.1-0.20190118093823-f849b5445de4/chain.go:34 +0x9c
go.etcd.io/etcd/etcdserver/api/v3rpc.newUnaryInterceptor.func1(0x57a4f20, 0xc002403e90, 0x54102a0, 0xc002403ec0, 0xc00282b5a0, 0xc002623ae0, 0xc001713858, 0x40897b6, 0x5ee878b1, 0x2fbc69c8)
	/Users/pingcap/gopkg/pkg/mod/go.etcd.io/etcd@v0.5.0-alpha.5.0.20191023171146-3cf2f69b5738/etcdserver/api/v3rpc/interceptor.go:64 +0x108
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1.1(0x57a4f20, 0xc002403e90, 0x54102a0, 0xc002403ec0, 0x203000, 0x203000, 0x4664bec, 0xc002403ec0)
	/Users/pingcap/gopkg/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.0.1-0.20190118093823-f849b5445de4/chain.go:34 +0x9c
go.etcd.io/etcd/etcdserver/api/v3rpc.newLogUnaryInterceptor.func1(0x57a4f20, 0xc002403e90, 0x54102a0, 0xc002403ec0, 0xc00282b5a0, 0xc002623ae0, 0x0, 0x0, 0x0, 0x0)
	/Users/pingcap/gopkg/pkg/mod/go.etcd.io/etcd@v0.5.0-alpha.5.0.20191023171146-3cf2f69b5738/etcdserver/api/v3rpc/interceptor.go:71 +0xba
github.com/grpc-ecosystem/go-grpc-middleware.ChainUnaryServer.func1(0x57a4f20, 0xc002403e90, 0x54102a0, 0xc002403ec0, 0xc00282b5a0, 0xc00282b5c0, 0xc001713a80, 0x4092ec8, 0x538efe0, 0xc002403e90)
	/Users/pingcap/gopkg/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.0.1-0.20190118093823-f849b5445de4/chain.go:39 +0x155
github.com/pingcap/kvproto/pkg/pdpb._PD_StoreHeartbeat_Handler(0x5499340, 0xc000473c20, 0x57a4f20, 0xc002403e90, 0xc0022349c0, 0xc0005dc750, 0x57a4f20, 0xc002403e90, 0xc0002cb540, 0x3e)
	/Users/pingcap/gopkg/pkg/mod/github.com/pingcap/kvproto@v0.0.0-20200409042513-05af14db7537/pkg/pdpb/pdpb.pb.go:4628 +0x14b
google.golang.org/grpc.(*Server).processUnaryRPC(0xc000001200, 0x57bc560, 0xc0011d7980, 0xc001820c00, 0xc0003956e0, 0x6ebdce8, 0x0, 0x0, 0x0)
	/Users/pingcap/gopkg/pkg/mod/google.golang.org/grpc@v1.25.1/server.go:1007 +0x460
google.golang.org/grpc.(*Server).handleStream(0xc000001200, 0x57bc560, 0xc0011d7980, 0xc001820c00, 0x0)
	/Users/pingcap/gopkg/pkg/mod/google.golang.org/grpc@v1.25.1/server.go:1287 +0xd97
google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc0003be2c0, 0xc000001200, 0x57bc560, 0xc0011d7980, 0xc001820c00)
	/Users/pingcap/gopkg/pkg/mod/google.golang.org/grpc@v1.25.1/server.go:722 +0xbb
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/Users/pingcap/gopkg/pkg/mod/google.golang.org/grpc@v1.25.1/server.go:720 +0xa1
Success to stop component `pd`
```

### What did you do?

start PD with config
```
[log]
level = "info"

[pd-server]
key-type = "txn"
use-region-storage = true

[replication]
enable-placement-rules = false
location-labels = ["zone", "rack"]
max-replicas = 1
strictly-match-label = false

[schedule]
enable-cross-table-merge = true
enable-one-way-merge = true
high-space-ratio = 0.5021591283248688
hot-region-cache-hits-threshold = 9
hot-region-schedule-limit = 13
leader-schedule-limit = 5
leader-schedule-policy = "count"
low-space-ratio = 0.9349351020978699
max-merge-region-keys = 20694
max-merge-region-size = 46
max-pending-peer-count = 1
max-snapshot-count = 3
max-store-down-time = "11m40s"
merge-schedule-limit = 24
patrol-region-interval = "968ms"
region-schedule-limit = 1723
replica-schedule-limit = 213
scheduler-max-waiting-operator = 3
split-merge-interval = "18m20s"
store-balance-rate = 29.19276264350798
store-limit-mode = "auto"
tolerant-size-ratio = 3.953397120203804
```
<!-- If possible, provide a recipe for reproducing the error. -->

### What did you expect to see?

### What did you see instead?

### What version of PD are you using (`pd-server -V`)?
```
[2020/06/16 15:52:16.767 +08:00] [INFO] [util.go:51] ["Welcome to Placement Driver (PD)"]
[2020/06/16 15:52:16.768 +08:00] [INFO] [util.go:52] [PD] [release-version=v4.0.0-rc-18-g8ccdf218d]
[2020/06/16 15:52:16.768 +08:00] [INFO] [util.go:53] [PD] [git-hash=8ccdf218db65e861bbdffc0a7e49a8e1865bad47]
[2020/06/16 15:52:16.768 +08:00] [INFO] [util.go:54] [PD] [git-branch=HEAD]
[2020/06/16 15:52:16.768 +08:00] [INFO] [util.go:55] [PD] [utc-build-time="2020-04-20 12:09:24"]
```

--- a/pkg/mock/mockoption/mockoption.go
+++ b/pkg/mock/mockoption/mockoption.go
@@ -43,6 +43,7 @@ const (
 	defaultLeaderSchedulePolicy        = "count"
 	defaultEnablePlacementRules        = false
 	defaultKeyType                     = "table"
+	defaultStoreLimitMode              = "manual"
 )
 
 // ScheduleOptions is a mock of ScheduleOptions
@@ -71,6 +72,7 @@ type ScheduleOptions struct {
 	TolerantSizeRatio            float64
 	LowSpaceRatio                float64
 	HighSpaceRatio               float64
+	StoreLimitMode               string
 	EnableRemoveDownReplica      bool
 	EnableReplaceOfflineReplica  bool
 	EnableMakeUpReplica          bool
@@ -110,6 +112,7 @@ func NewScheduleOptions() *ScheduleOptions {
 	mso.TolerantSizeRatio = defaultTolerantSizeRatio
 	mso.LowSpaceRatio = defaultLowSpaceRatio
 	mso.HighSpaceRatio = defaultHighSpaceRatio
+	mso.StoreLimitMode = defaultStoreLimitMode
 	mso.EnableRemoveDownReplica = true
 	mso.EnableReplaceOfflineReplica = true
 	mso.EnableMakeUpReplica = true
@@ -278,4 +281,9 @@ func (mso *ScheduleOptions) GetLeaderSchedulePolicy() core.SchedulePolicy {
 // GetKeyType is to get key type.
 func (mso *ScheduleOptions) GetKeyType() core.KeyType {
 	return core.StringToKeyType(mso.KeyType)
+}
+
+// GetStoreLimitMode returns the limit mode of store.
+func (mso *ScheduleOptions) GetStoreLimitMode() string {
+	return mso.StoreLimitMode
 }

--- a/server/schedule/operator_controller.go
+++ b/server/schedule/operator_controller.go
@@ -897,7 +897,8 @@ func (oc *OperatorController) SetAllStoresLimitAuto(rate float64, limitType stor
 	for _, s := range stores {
 		sid := s.GetID()
 		if old, ok := oc.storesLimit[sid]; ok {
-			if old[limitType].Mode() == storelimit.Manual {
+			limit, ok1 := old[limitType]
+			if ok1 && limit.Mode() == storelimit.Manual {
 				continue
 			}
 		}

--- a/server/schedule/operator_controller_test.go
+++ b/server/schedule/operator_controller_test.go
@@ -704,3 +704,15 @@ func (t *testOperatorControllerSuite) TestAddWaitingOperator(c *C) {
 	// no space left, new operator can not be added.
 	c.Assert(controller.AddWaitingOperator(addPeerOp(0)), Equals, 0)
 }
+
+func (t *testOperatorControllerSuite) TestAutoStoreLimitMode(c *C) {
+	opt := mockoption.NewScheduleOptions()
+	opt.StoreLimitMode = "auto"
+	tc := mockcluster.NewCluster(opt)
+	stream := mockhbstream.NewHeartbeatStreams(tc.ID, true /* no need to run */)
+	oc := NewOperatorController(t.ctx, tc, stream)
+
+	tc.AddLeaderStore(1, 10)
+	oc.SetStoreLimit(1, 10, storelimit.Auto, storelimit.RegionAdd)
+	oc.SetAllStoresLimitAuto(10, storelimit.RegionRemove)
+}

